core/src/main/java/org/apache/calcite/statistic/MapSqlStatisticProvider.java

```diff
@@ -0,0 +1,138 @@
+        .put("foodmart", "region", 24d, "id")
+        .put("foodmart", "salary", 21_252d, "id")
+        .put("foodmart", "sales_fact_1997", 86_837d, "id")
+        .put("foodmart", "store", 25d, "id")
+        .put("foodmart", "store_ragged", 25d, "id")
+        .put("foodmart", "time_by_day", 730d, "id")
+        .put("foodmart", "warehouse", 24d, "id")
+        .put("foodmart", "warehouse_class", 6d, "id")
+        .put("scott", "EMP", 10d, "id")
+        .put("scott", "DEPT", 4d, "id")
+        .put("tpcds", "CALL_CENTER", 8d, "id")
+        .put("tpcds", "CATALOG_PAGE", 11_718d, "id")
+        .put("tpcds", "CATALOG_RETURNS", 144_067d, "id")
+        .put("tpcds", "CATALOG_SALES", 1_441_548d, "id")
+        .put("tpcds", "CUSTOMER", 100_000d, "id")
+        .put("tpcds", "CUSTOMER_ADDRESS", 50_000d, "id")
+        .put("tpcds", "CUSTOMER_DEMOGRAPHICS", 1_920_800d, "id")
+        .put("tpcds", "DATE_DIM", 73049d, "id")
+        .put("tpcds", "DBGEN_VERSION", 1d, "id")
+        .put("tpcds", "HOUSEHOLD_DEMOGRAPHICS", 7200d, "id")
+        .put("tpcds", "INCOME_BAND", 20d, "id")
+        .put("tpcds", "INVENTORY", 11_745_000d, "id")
+        .put("tpcds", "ITEM", 18_000d, "id")
+        .put("tpcds", "PROMOTION", 300d, "id")
+        .put("tpcds", "REASON", 35d, "id")
+        .put("tpcds", "SHIP_MODE", 20d, "id")
+        .put("tpcds", "STORE", 12d, "id")
+        .put("tpcds", "STORE_RETURNS", 287_514d, "id")
+        .put("tpcds", "STORE_SALES", 2_880_404d, "id")
+        .put("tpcds", "TIME_DIM", 86_400d, "id")
+        .put("tpcds", "WAREHOUSE", 5d, "id")
+        .put("tpcds", "WEB_PAGE", 60d, "id")
+        .put("tpcds", "WEB_RETURNS", 71_763d, "id")
+        .put("tpcds", "WEB_SALES", 719_384d, "id")
+        .put("tpcds", "WEB_SITE", 1d, "id");
+    cardinalityMap = initializer.cardinalityMapBuilder.build();
+    keyMap = initializer.keyMapBuilder.build();
+  }
+
+  public double tableCardinality(RelOptTable table) {
+    final JdbcTable jdbcTable = table.unwrap(JdbcTable.class);
+    final List<String> qualifiedName;
+    if (jdbcTable != null) {
+      qualifiedName = Arrays.asList(jdbcTable.jdbcSchemaName,
+          jdbcTable.jdbcTableName);
+    } else {
+      qualifiedName = table.getQualifiedName();
+    }
+    return cardinalityMap.get(qualifiedName.toString());
+  }
+
+  public boolean isForeignKey(RelOptTable fromTable, List<Integer> fromColumns,
+      RelOptTable toTable, List<Integer> toColumns) {
+    // Assume that anything that references a primary key is a foreign key.
+    // It's wrong but it's enough for our current test cases.
+    return isKey(toTable, toColumns);
+  }
+
+  public boolean isKey(RelOptTable table, List<Integer> columns) {
+    return columns.size() == 1
+        && keyMap.get(table.getQualifiedName().toString())
+            .equals(
+                ImmutableList.of(table.getRowType().getFieldList()
+                  .get(columns.get(0)).getName()));
```

Comment:
We should pre-check for indexOutofBounds for ImmutableList.of(table.getRowType().getFieldList().get(columns.get(0))

Reply:
I changed it to fail the test if any of the columns are out of range. Also it now allows composite keys (even though none of those currently defined are composite).